pkg/api/controller.go

```diff
@@ -263,6 +263,166 @@ func (c *Controller) CreatePresignMultipartUpload(w http.ResponseWriter, r *http
 	writeResponse(w, r, http.StatusCreated, resp)
 }
 
+func (c *Controller) UploadPart(w http.ResponseWriter, r *http.Request, body apigen.UploadPartJSONRequestBody, dstRepository string, branch string, uploadID string, partNumber int, params apigen.UploadPartParams) {
+	var (
+		ctx     = r.Context()
+		dstPath = params.Path
+	)
+	requiredPermissions := permissions.Node{
+		Permission: permissions.Permission{
+			Action:   permissions.WriteObjectAction,
+			Resource: permissions.ObjectArn(dstRepository, dstPath),
+		},
+	}
+	if !c.authorize(w, r, requiredPermissions) {
+		return
+	}
+
+	c.LogAction(ctx, "upload_part", r, dstRepository, dstPath, "")
+
+	repo, err := c.Catalog.GetRepository(ctx, dstRepository)
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	// verify physical address
+	physicalAddress, addressType := normalizePhysicalAddress(repo.StorageNamespace, body.PhysicalAddress)
+	if addressType != catalog.AddressTypeRelative {
+		writeError(w, r, http.StatusBadRequest, "physical address must be relative to the storage namespace")
+		return
+	}
+
+	//  verify it has been saved for linking
+	if err := c.Catalog.VerifyLinkAddress(dstRepository, branch, params.Path, physicalAddress); c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	dstObjectRef := block.ObjectPointer{
+		StorageID:        repo.StorageID,
+		StorageNamespace: repo.StorageNamespace,
+		IdentifierType:   block.IdentifierTypeRelative,
+		Identifier:       physicalAddress,
+	}
+
+	presignedURL, err := c.BlockAdapter.GetPresignUploadPartURL(ctx, dstObjectRef, uploadID, partNumber)
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+
+	response := apigen.UploadTo{
+		PresignedUrl: presignedURL,
+	}
+	writeResponse(w, r, http.StatusOK, response)
+}
+
+func (c *Controller) UploadPartCopy(w http.ResponseWriter, r *http.Request,
+	body apigen.UploadPartCopyJSONRequestBody, dstRepository string, branch string, uploadID string, partNumber int,
+	params apigen.UploadPartCopyParams) {
+	var (
+		ctx = r.Context()
+
+		dstPath = params.Path
+
+		srcRepository = body.CopySource.Repository
+		srcRef        = body.CopySource.Ref
+		srcPath       = body.CopySource.Path
+	)
+
+	requiredPermissions := permissions.Node{
+		Type: permissions.NodeTypeAnd,
+		Nodes: []permissions.Node{
+			{
+				Permission: permissions.Permission{
+					Action:   permissions.WriteObjectAction,
+					Resource: permissions.ObjectArn(dstRepository, dstPath),
+				},
+			},
+			{
+				Permission: permissions.Permission{
+					Action:   permissions.ReadObjectAction,
+					Resource: permissions.ObjectArn(srcRepository, srcPath),
+				},
+			},
+		},
+	}
+	if !c.authorize(w, r, requiredPermissions) {
+		return
+	}
+
+	c.LogAction(ctx, "upload_part_copy", r, dstRepository, dstPath, srcPath)
```

Comment:
notify BI about this type in case there are any analytics that track specific usage.

Reply:
Done (off review)